src/ring_theory/localization.lean

```diff
@@ -1390,6 +1393,102 @@ def localization_localization_at_prime_iso_localization (p : ideal (localization
   localization.at_prime (p.comap (algebra_map R _)) ≃ₐ[R] localization.at_prime p :=
 is_localization.alg_equiv (p.comap (algebra_map R _)).prime_compl _ _
 
+end
+
+variables (S)
+
+/-- If `M ≤ N` are submonoids of `R`, then there is a natural map `M⁻¹S →+* N⁻¹S`. -/
+noncomputable
+def localization_algebra_of_submonoid_le
+  (M N : submonoid R) (h : M ≤ N) [is_localization M S] [is_localization N T] :
+  algebra S T :=
+begin
+  apply ring_hom.to_algebra,
+  apply @is_localization.lift _ _ M _ _ _ _ _ _ (algebra_map _ _),
+  any_goals { apply_instance },
+  intro y,
+  apply is_localization.map_units T ⟨_, h y.prop⟩,
+end
+
+/-- If `M ≤ N` are submonoids of `R`, then the natural map `M⁻¹S →+* N⁻¹S` commutes with the
+localization maps -/
+lemma localization_is_scalar_tower_of_submonoid_le
+  (M N : submonoid R) (h : M ≤ N) [is_localization M S] [is_localization N T] :
+  @@is_scalar_tower R S T _ (localization_algebra_of_submonoid_le S T M N h).to_has_scalar _ :=
+begin
+  letI := localization_algebra_of_submonoid_le S T M N h,
+  exact is_scalar_tower.of_algebra_map_eq' (is_localization.lift_comp _).symm
+end
+
+noncomputable
+instance (x : ideal R) [H : x.is_prime] [is_domain R] :
+  algebra (localization.at_prime x) (localization (non_zero_divisors R)) :=
+localization_algebra_of_submonoid_le _ _ x.prime_compl (non_zero_divisors R)
+  (by { intros a ha, rw mem_non_zero_divisors_iff_ne_zero, exact λ h, ha (h.symm ▸ x.zero_mem) })
+
+/-- If `M ≤ N` are submonoids of `R`, then `N⁻¹S` is also the localization of `M⁻¹S` at `N`. -/
+lemma is_localization_of_submonoid_le
+  (M N : submonoid R) (h : M ≤ N) [is_localization M S] [is_localization N T] :
+  @@is_localization _ (N.map (algebra_map R S).to_monoid_hom) T _
+    (localization_algebra_of_submonoid_le S T M N h) :=
```

Comment:
This lemma is really painful, both in its statement and its proof, but unfortunately I don't know this part of mathlib well enough to help.

Reply:
@Vierkantor do you have an idea how to improve this?

Reply:
Hmm, we can't just make the definitions above instances by turning the hypothesis into `[fact (M ≤ N)]`, since `M` nor `N` is obvious from context here. An alternative is to take `[algebra S T]` as a parameter + the equivalent of the `letI := is_scalar_tower` as an explicit hypothesis, then whenever such an algebra is not already available, we can create it with `localization_algebra_of_submonoid_le`.